### PR TITLE
Retry awaiting for stop() in background_asyncio_service()

### PR DESCRIPTION
**Pull request by gsalgado**
In trinity, we inject a KeyboardInterrupt into a couroutine to tell it to stop,
and it will bubble up to background_asyncio_service(), whose finally: block
awaits for manager.stop(). However, for some of our services the manager.stop()
raises yet another KeyboardInterrupt, preventing trinity from shutting
down. I couldn't figure out where this 2nd KeyboardInterrupt comes from,
but it seems to happen always on the same services, and retrying that
always works (as in, the stop() call returns without error)

This is another of the obscure behaviours originally reported in
https://github.com/ethereum/trinity/issues/1613

### CODE REVIEW

**Comment by gsalgado** [comment]
This is what I see on the logs when shutting down trinity with this change

```
^C WARNING  2020-04-09 18:42:07,491               Manager  UPnPService: Raised KeyboardInterrupt!                                                                                                                  
 WARNING  2020-04-09 18:42:07,492               Manager  IPCServer: Raised KeyboardInterrupt!                                                                                                                      
 WARNING  2020-04-09 18:42:07,493               Manager  PeerDBServer: Raised KeyboardInterrupt!                   
 WARNING  2020-04-09 18:42:07,950               Manager  AsyncioEventBusService: no exception raised
 WARNING  2020-04-09 18:42:07,942               Manager  AsyncioEventBusService: no exception raised
    INFO  2020-04-09 18:42:07,951                TxPool  Stopping Tx Pool...
 WARNING  2020-04-09 18:42:07,953               Manager  AsyncioEventBusService: no exception raised
 WARNING  2020-04-09 18:42:07,940               Manager  FullNode: Raised KeyboardInterrupt!
 WARNING  2020-04-09 18:42:07,962               Manager  AsyncioEventBusService: no exception raised
 WARNING  2020-04-09 18:42:07,960               Manager  UPnPService: Got KeyboardInterrupt while waiting for stop(), trying again
 WARNING  2020-04-09 18:42:07,947               Manager  AsyncioEventBusService: no exception raised
 WARNING  2020-04-09 18:42:07,948               Manager  PeerDBServer: Got KeyboardInterrupt while waiting for stop(), trying again
 WARNING  2020-04-09 18:42:07,947               Manager  IPCServer: Got KeyboardInterrupt while waiting for stop(), trying again
 WARNING  2020-04-09 18:42:07,947               Manager  IPCServer: stopped!
 WARNING  2020-04-09 18:42:07,950               Manager  AsyncioEventBusService: no exception raised
 WARNING  2020-04-09 18:42:07,977               Manager  AsyncioEventBusService: no exception raised
 WARNING  2020-04-09 18:42:07,960               Manager  UPnPService: stopped!
 WARNING  2020-04-09 18:42:07,960               Manager  AsyncioEventBusService: no exception raised
 WARNING  2020-04-09 18:42:07,989               Manager  AsyncioEventBusService: no exception raised
 WARNING  2020-04-09 18:42:07,948               Manager  PeerDBServer: stopped!
 WARNING  2020-04-09 18:42:07,949               Manager  ConnectionTrackerServer: Raised KeyboardInterrupt!
 WARNING  2020-04-09 18:42:07,957               Manager  ConnectionTrackerServer: stopped!
 WARNING  2020-04-09 18:42:07,957               Manager  AsyncioEventBusService: no exception raised
 WARNING  2020-04-09 18:42:07,984               Manager  AsyncioEventBusService: stopped!
    INFO  2020-04-09 18:42:08,026            FullServer  Closing server...
```

File: async_service/asyncio.py [hunk 1]
```diff
@@ -413,7 +413,24 @@ def external_api(func: TFunc) -> TFunc:
...
+                else:
+                    manager.logger.exception("%s: Raised other exception!", service)
+                try:
+                    await manager.stop()
+                    manager.logger.warning("%s: stopped!", service)
+                except KeyboardInterrupt:
+                    manager.logger.warning(
+                        "%s: Got KeyboardInterrupt while waiting for stop(), trying again", service)
+                    await manager.stop()
+                    manager.logger.warning("%s: stopped!", service)
```
Comment (pipermerriam) [inline async_service/asyncio.py:430]: seems like the exception should be re-raised after the stop call.

**Review by pipermerriam — COMMENTED** [review]
I'm nervious about this.  Specifically, at the point where we're handling a `KeyboardInterrupt` I think the shutdown situation is more *dire* and maybe justifies a more forceful shutdown.

`trio` has an API that I'm drawing inspiration from: [`AsyncResource.aclose_forcefully`](https://trio.readthedocs.io/en/stable/reference-io.html#trio.aclose_forcefully).  The current `stop` mechanics are still subject to unwinding all the various tasks in the proper order.  I'm thinking that we could implement a `stop_forcefully` that does a more aggressive cancellation, immediately and concurrently cancelling the entire task DAG.

**Comment by gsalgado** [comment]
Found the reason [why the double KeyboardInterrupt is happening](https://github.com/ethereum/trinity/issues/1613#issuecomment-611812510), so we can fix that instead of resorting to this hack

**Pull request closed without merging**